### PR TITLE
[IOTDB-2899] Feature/set storage group

### PR DESCRIPTION
**Pull request by SpriCoder**
See more details: https://issues.apache.org/jira/browse/IOTDB-2899

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/parser/ASTVisitor.java [hunk 1]
```diff
@@ -1564,6 +1564,11 @@ private String[] parsePrivilege(IoTDBSqlParser.PrivilegesContext ctx) {
     return privileges.toArray(new String[0]);
   }
 
+  @Override
+  public Statement visitSetStorageGroup(IoTDBSqlParser.SetStorageGroupContext ctx) {
+    return super.visitSetStorageGroup(ctx);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/parser/ASTVisitor.java:1569]: You should construct SetStorageGroupStatement here by yourselft. You can refer the implementation in `visitSetStorageGroup` of `IoTDBSqlVisitor`.

File: server/src/main/java/org/apache/iotdb/db/mpp/sql/analyze/Analyzer.java [hunk 2]
```diff
@@ -230,6 +231,16 @@ public Analysis visitAlterTimeseries(
...
+    @Override
+    public Analysis visitSetStorageGroup(
+        SetStorageGroupStatement setStorageGroupStatement, MPPQueryContext context) {
+      Analysis analysis = new Analysis();
+      analysis.setStatement(setStorageGroupStatement);
+
+      // TODO SpriCoder check what to do
+      return analysis;
+    }
+
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/sql/analyze/Analyzer.java:243]: You don't need to do analyze here, Config related Statement won't reach here.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/IConfigTask.java [hunk 3]
```diff
@@ -17,10 +17,13 @@
...
+package org.apache.iotdb.db.mpp.execution.config;
+
+import org.apache.iotdb.db.exception.metadata.MetadataException;
+import org.apache.iotdb.rpc.IoTDBConnectionException;
 
 import com.google.common.util.concurrent.ListenableFuture;
 
 public interface IConfigTask {
-  ListenableFuture<Void> execute();
+  ListenableFuture<Void> execute() throws IoTDBConnectionException, MetadataException;
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/IConfigTask.java:28]: You don' have to throw these checked execption here.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    // Get response or throw exception
+    TSStatus tsStatus = configNodeClient.setStorageGroup(req);
+    if (TSStatusCode.SUCCESS_STATUS.getStatusCode() != tsStatus.getCode()) {
+      throw new MetadataException("Failed to set storage Group");
+    }
+    // If the action is executed successfully, return the Future.
+    // If your operation is async, you can return the corresponding future directly.
+    return Futures.immediateVoidFuture();
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:71]:
```
    SettableFuture<Void> future = SettableFuture.create();
    try {
      // do something you want to do
      future.set(null);
    } catch (IoTDBConnectionException | BadNodeUrlException e) {
      // catch checked exception here.
      future.setException(e);
    }
    return future;
```

Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:71]: Avoid throw checked exception out of this scope.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Comment by SpriCoder** [comment]
Thanks for you review. all problem are already fixed. @JackieTien97 PTAL!

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SetStorageGroupTask implements IConfigTask {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SetStorageGroupTask.class);
+  private Statement statement;
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:40]:
```suggestion
  private final SetStorageGroupStatement statement;
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  public SetStorageGroupTask(Statement statement) {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:42]:
```suggestion
  public SetStorageGroupTask(SetStorageGroupStatement statement) {
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    this.statement = statement;
+  }
+
+  @Override
+  public ListenableFuture<Void> execute() {
+    SettableFuture<Void> future = SettableFuture.create();
+    if (!(statement instanceof SetStorageGroupStatement)) {
+      LOGGER.error("SetStorageGroup not get SetStorageGroupStatement");
+      return Futures.immediateVoidFuture();
+    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:52]: Don't need to do this if judgement here.

Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:52]:
```suggestion
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    ConfigNodeClient configNodeClient = null;
+    try {
+      configNodeClient = new ConfigNodeClient();
+      // Send request to some API server
+      TSStatus tsStatus = configNodeClient.setStorageGroup(req);
+      // Get response or throw exception
+      if (TSStatusCode.SUCCESS_STATUS.getStatusCode() != tsStatus.getCode()) {
+        LOGGER.error("Failed to connect to config node.");
+        future.setException(new MetadataException("Failed to connect to config node."));
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:66]:
```suggestion
        LOGGER.error("Failed to execute set storage group {} in config node, status is {}.", statement.getStorageGroupPath(), tsStatus);
        future.setException(new StatementExecutionException(tsStatus));
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/config/SetStorageGroupTask.java:67]:
```suggestion
      } else {
        future.set(null);
      }
```

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**